src/libos/src/fs/file_ops/dirent.rs

```diff
@@ -1,19 +1,99 @@
 use super::*;
 
+pub fn do_getdents64(fd: FileDesc, buf: &mut [u8]) -> Result<usize> {
+    getdents_common::<u8>(fd, buf)
+}
+
+pub fn do_getdents(fd: FileDesc, buf: &mut [u8]) -> Result<usize> {
+    getdents_common::<()>(fd, buf)
+}
+
+fn getdents_common<T: DirentType + Copy + Default>(fd: FileDesc, buf: &mut [u8]) -> Result<usize> {
+    debug!(
+        "getdents: fd: {}, buf: {:?}, buf_size: {}",
+        fd,
+        buf.as_ptr(),
+        buf.len(),
+    );
+
+    let file_ref = current!().file(fd)?;
+    let info = file_ref.metadata()?;
+    if info.type_ != FileType::Dir {
+        return_errno!(ENOTDIR, "");
+    }
+    let mut writer = unsafe { DirentBufWriter::new(buf) };
+    loop {
+        let name = match file_ref.read_entry() {
+            Err(e) => {
+                let errno = e.errno();
+                if errno == ENOENT {
+                    break;
+                }
+                return Err(e.cause_err(|_| errno!(errno, "failed to read entry")));
+            }
+            Ok(name) => name,
+        };
+        // TODO: get ino from dirent
+        let dirent = LinuxDirent::<T>::new(1, &name);
+        if let Err(e) = writer.try_write(&dirent, &name) {
+            file_ref.seek(SeekFrom::Current(-1))?;
+            if writer.written_size == 0 {
+                return Err(e);
+            } else {
+                break;
+            }
+        }
+    }
+    Ok(writer.written_size)
+}
+
 #[repr(packed)] // Don't use 'C'. Or its size will align up to 8 bytes.
-struct LinuxDirent64 {
+struct LinuxDirent<T: DirentType + Copy + Default> {
     /// Inode number
     ino: u64,
     /// Offset to next structure
     offset: u64,
     /// Size of this dirent
     reclen: u16,
     /// File type
-    type_: u8,
-    /// Filename (null-terminated)
-    name: [u8; 0],
```

Comment:
I suggest keeping the `name` field, although it is zero length and thus "useless" in practice. But it can serve as a document to remind us that `LinuxDirent` actually has such a field.

Reply:
Ok